### PR TITLE
Add libcare-stresstest.

### PR DESCRIPTION
**Pull request by rashchupkinr**
libcare-stresstest as separate binary.

### CODE REVIEW

File: src/kpatch_user.c [hunk 1]
```diff
@@ -25,9 +25,13 @@
 #include "list.h"
 #include "kpatch_log.h"
 
+
 /* Global variables */
 static char storage_dir[PATH_MAX] = "/var/lib/libcare";
-
+struct test_data {
```
Comment (paboldin) [inline src/kpatch_user.c]: If this is only used under STRESS_TEST please move that declaration bellow.

File: src/kpatch_user.c [hunk 2]
```diff
@@ -589,6 +593,93 @@ cmd_update(int argc, char *argv[])
...
+
+	rv = sscanf(argv[1], "%d", &pid);
+	if (rv != 1) {
+		kperr("can't parse pid from %s", argv[1]);
+		return -1;
+	}
+	sprintf(pid_str, "%d", pid);
+
+	optind = 1;
+	rv = cmd_unpatch_user(ARRAY_SIZE(patch_pid_argv_unpatch),
```
Comment (paboldin) [inline src/kpatch_user.c]: I was wrong when picked up this approach.

Let's do it otherwise. Let's use `processes_unpatch` here. I will fix other code as well.

File: src/kpatch_user.c [hunk 3]
```diff
@@ -589,6 +593,93 @@ cmd_update(int argc, char *argv[])
...
+{
+	kpatch_process_t _proc, *proc = &_proc;
+	struct timespec req, rem;
+	int rv, i;
+	req.tv_sec = 0;
+	req.tv_nsec = 1000*1000;
+	for (i=0; i<period; i++) {
+		nanosleep(&req, &rem);
+		rv = kpatch_process_init(proc, pid, 0, -1);
+		close(proc->fdmaps);
```
Comment (paboldin) [inline src/kpatch_user.c]: Although it is not required here please use `kpatch_process_free`. This makes API usage consistent.

File: src/kpatch_user.c [hunk 4]
```diff
@@ -589,6 +593,93 @@ cmd_update(int argc, char *argv[])
...
+	if (rv != 1) {
+		kperr("can't parse pid from %s", argv[1]);
+		return -1;
+	}
+
+	while (1) {
+		period = rand() % test_info.option_period;
+		if (server_wait(pid, period) < 0)
+			return 0;
+		if ((rv = cmd_execve_startup(fd, argc, argv, 0)) < 0) {
```
Comment (paboldin) [inline src/kpatch_user.c]: While it should work with fd=-1 I think we should call the patch_user as it never sets `-s` argument which indicates that the process was just started. Also, we can hide away all the `-r` and `-s` options since they are not needed anymore.

File: src/kpatch_user.c [hunk 5]
```diff
@@ -598,8 +689,16 @@ server_execute_cmd(int fd, int argc, char *argv[])
 
 	if (!strcmp(cmd, "execve"))
 		return cmd_execve_startup(fd, argc, argv, 1);
-	if (!strcmp(cmd, "startup"))
+	if (!strcmp(cmd, "startup")) {
+#ifdef STRESS_TEST
+		if (test_info.option_period != 0)
+			return server_stress_test(fd, argc, argv);
```
Comment (paboldin) [inline src/kpatch_user.c]: Since we have a separate binary we can just call `server_stress_test` here.

File: src/kpatch_user.c [hunk 6]
```diff
@@ -765,6 +863,10 @@ cmd_server(int argc, char *argv[])
 		return -1;
 	}
 
+#ifdef STRESS_TEST
+	sscanf(argv[0], "%d", &test_info.option_period);
```
Comment (paboldin) [inline src/kpatch_user.c]: Please check the `sscanf` return value.

File: src/kpatch_user.c [hunk 7]
```diff
@@ -882,7 +984,6 @@ execute_cmd(int argc, char *argv[])
 int main(int argc, char *argv[])
 {
 	int opt;
-
```
Comment (paboldin) [inline src/kpatch_user.c]: Please avoid unnecessary edits. I like lines after variable declarations, let's keep this one :)

**Review by paboldin — CHANGES_REQUESTED** [review]

File: src/kpatch_coro.c [hunk 8]
```diff
@@ -622,7 +622,8 @@ void kpatch_coroutines_free(struct kpatch_process *proc)
 	if (proc->coro.unwd)
 		unw_destroy_addr_space(proc->coro.unwd);
 
-	list_for_each_entry_safe(c, tmp, &proc->coro.coros, list) {
-		kpatch_coro_free(c);
-	}
+	if (!list_empty(&proc->coro.coros))
```
Comment (paboldin) [inline src/kpatch_coro.c:625]: Why is this required? Iterating empty list is the same as this.

File: src/kpatch_process.c [hunk 9]
```diff
@@ -1121,10 +1121,10 @@ kpatch_process_init(kpatch_process_t *proc,
 	list_init(&proc->vmaholes);
 	proc->num_objs = 0;
 
-	if (process_get_comm(proc))
-		goto out_unlock;
 	if (kpatch_coroutines_init(proc))
 		goto out_unlock;
+	if (process_get_comm(proc))
```
Comment (paboldin) [inline src/kpatch_process.c:1126]: May you please elaborate on this?

Comment (rashchupkinr) [reply src/kpatch_process.c:1126]: If process already exited, process_get_comm returns error, proc->coro.coros list is not correctly initialized and process exits on exception at list iteration.

File: src/kpatch_user.c [hunk 10]
```diff
@@ -563,6 +564,87 @@ cmd_update(int argc, char *argv[])
...
+
+static int cmd_stress_test(int fd, int argc, char *argv[])
+{
+	int child = fork();
+	if (child == 0) {
+		int rv = server_stress_test(argc, argv);
+		exit(rv);
+	}
+
+	int reply=0;
```
Comment (paboldin) [inline src/kpatch_user.c]: Why is this required? The child should take care of it, the parent should close the fd.

**Review by paboldin — CHANGES_REQUESTED** [review]

**Review by paboldin — APPROVED** [review]
